File: lib/premium-analysis/agents-outputs.ts
import {
  buildChargeDescriptionLookup,
  shipmentPackageDedupeKey,
  type ChargeDescriptionMappingRow,
  type InvoiceAnalysisSummary,
} from '@/lib/premium-analysis/analysis-summary'
import type { AgentsAnalysisExtensions } from '@/lib/premium-analysis/agents-types'
import { detectAnomalies, buildDatasetFlags } from '@/lib/premium-analysis/anomaly-detection'
import { prioritizeActions } from '@/lib/premium-analysis/action-prioritization'
import { buildCarrierMix } from '@/lib/premium-analysis/carrier-mix'
import {
  buildShipmentFacts,
  shipmentWeightGapLbs,
} from '@/lib/premium-analysis/shipment-fact'
import {
  detectContractDiscountShortfalls,
  type ContractDiscounts,
} from '@/lib/premium-analysis/contract-compliance'
import { estimateSavings } from '@/lib/premium-analysis/savings-estimator'
import { rollupByAgentsCategory } from '@/lib/premium-analysis/spec-categories'
import type { InvoiceRecord } from '@/lib/invoices/csv'

export function enrichSummaryWithAgentsOutputs(
  summary: InvoiceAnalysisSummary,
  records: InvoiceRecord[],
  mappingRows: ChargeDescriptionMappingRow[] | null | undefined,
  contractDiscounts: ContractDiscounts = {}
): InvoiceAnalysisSummary & AgentsAnalysisExtensions {
  const mappingLookup = buildChargeDescriptionLookup(mappingRows)
  const shipmentFacts = buildShipmentFacts(records, mappingLookup, mappingRows)
  const specCategories = rollupByAgentsCategory(records, mappingLookup, mappingRows)
  const baseFreight = specCategories.categories.find((c) => c.category === 'BASE_FREIGHT')?.totalCost ?? 0

  const measures = {
    ...summary.measures,
    baseFreightCost: baseFreight,
    accessorialRate: baseFreight > 0 ? (summary.measures.costAccessorials ?? 0) / baseFreight : 0,
    weightGap: shipmentWeightGapLbs(shipmentFacts),
  }

  const shipmentCountsByCarrier = countShipmentsByDimension(records, 'carrier')
  const shipmentCountsByService = countShipmentsByDimension(records, 'service')

  const byCarrier = Object.fromEntries(
    Object.entries(summary.byCarrier).map(([k, v]) => [
      k,
      { ...v, shipmentCount: shipmentCountsByCarrier.get(k) ?? 0 },
    ])
  )
  const byService = Object.fromEntries(
    Object.entries(summary.byService).map(([k, v]) => [
      k,
      { ...v, shipmentCount: shipmentCountsByService.get(k) ?? 0 },
    ])
  )

  const enrichedBase: InvoiceAnalysisSummary = {
    ...summary,
    measures,
    byCarrier,
    byService,
  }

  const datasetFlags = buildDatasetFlags(enrichedBase, records, mappingLookup, mappingRows)
  let anomalyFlags = detectAnomalies(records, enrichedBase, mappingLookup, mappingRows, shipmentFacts)

  const contractFlags = detectContractDiscountShortfalls(records, contractDiscounts)
  anomalyFlags = [...anomalyFlags, ...contractFlags].sort((a, b) => b.amount - a.amount)

  const savingsEstimate = estimateSavings(
    anomalyFlags,
    summary.monthlySpend,
    summary.measures.totalCost
  )
  const actionItems = prioritizeActions(savingsEstimate)

  return {
    ...enrichedBase,
    specCategories,
    carrierMix: buildCarrierMix(shipmentFacts),
    anomalyFlags,
    savingsEstimate,
    actionItems,
    datasetFlags,
  }
}

function countShipmentsByDimension(
  records: InvoiceRecord[],
  dim: 'carrier' | 'service'
): Map<string, number> {
  const sets = new Map<string, Set<string>>()

  for (const rec of records) {
    const key =
      dim === 'carrier'
        ? (rec['Carrier Name'] ?? '').trim() || 'Unknown'
        : (rec['Original Service Description'] ?? '').trim() ||
          (rec['Charge Category Code'] ?? '').trim() ||
          'Unknown'
    const shipKey = shipmentPackageDedupeKey(rec)
    if (!shipKey) continue
    const bucket = sets.get(key) ?? new Set<string>()
    bucket.add(shipKey)
    sets.set(key, bucket)
  }

  return new Map([...sets.entries()].map(([k, v]) => [k, v.size]))
}

File: scripts/premium_analysis_mirror/agents_outputs.py
"""AGENTS outputs layer — mirrors lib/premium-analysis/agents-outputs.ts."""

from __future__ import annotations

from typing import Any

from .action_prioritization import prioritize_actions
from .anomaly_detection import build_dataset_flags, detect_anomalies
from .carrier_mix import build_carrier_mix
from .ingest_quality import apply_ingest_quality_gate, evaluate_ingest_quality
from .savings_estimator import estimate_savings
from .shipment_fact import build_shipment_facts, shipment_weight_gap_lbs
from .spec_categories import rollup_by_agents_category
from .primitives import shipment_package_dedupe_key


def _count_shipments_by_dimension(
    records: list[dict[str, Any]], dim: str
) -> dict[str, int]:
    sets: dict[str, set[str]] = {}
    for rec in records:
        if dim == "carrier":
            key = str(rec.get("Carrier Name") or "").strip() or "Unknown"
        else:
            key = (
                str(rec.get("Original Service Description") or "").strip()
                or str(rec.get("Charge Category Code") or "").strip()
                or "Unknown"
            )
        ship_key = shipment_package_dedupe_key(rec)
        if not ship_key:
            continue
        bucket = sets.setdefault(key, set())
        bucket.add(ship_key)
    return {k: len(v) for k, v in sets.items()}


def enrich_summary_with_agents_outputs(
    summary: dict[str, Any],
    records: list[dict[str, Any]],
    mapping_rows: list[dict[str, Any]] | None,
    mapping_lookup: dict[str, dict[str, str]],
    ingest_diagnostics: dict[str, Any] | None = None,
) -> dict[str, Any]:
    shipment_facts = build_shipment_facts(records, mapping_lookup, mapping_rows)
    spec_categories = rollup_by_agents_category(records, mapping_lookup, mapping_rows)
    base_freight = next(
        (c["totalCost"] for c in spec_categories["categories"] if c["category"] == "BASE_FREIGHT"),
        0.0,
    )

    measures = {
        **summary["measures"],
        "baseFreightCost": base_freight,
        "accessorialRate": (
            summary["measures"]["costAccessorials"] / base_freight if base_freight > 0 else 0.0
        ),
        "weightGap": shipment_weight_gap_lbs(shipment_facts),
    }

    ship_by_carrier = _count_shipments_by_dimension(records, "carrier")
    ship_by_service = _count_shipments_by_dimension(records, "service")

    by_carrier = {
        k: {**v, "shipmentCount": ship_by_carrier.get(k, 0)}
        for k, v in summary.get("byCarrier", {}).items()
    }
    by_service = {
        k: {**v, "shipmentCount": ship_by_service.get(k, 0)}
        for k, v in summary.get("byService", {}).items()
    }

    enriched_base = {**summary, "measures": measures, "byCarrier": by_carrier, "byService": by_service}

    dataset_flags = build_dataset_flags(enriched_base, records, mapping_lookup, mapping_rows)
    anomaly_flags = detect_anomalies(
        records, enriched_base, mapping_lookup, mapping_rows, shipment_facts
    )

    savings_estimate = estimate_savings(
        anomaly_flags,
        summary.get("monthlySpend") or [],
        summary["measures"]["totalCost"],
    )
    action_items = prioritize_actions(savings_estimate)

    result = {
        **enriched_base,
        "specCategories": spec_categories,
        "carrierMix": build_carrier_mix(shipment_facts),
        "anomalyFlags": anomaly_flags,
        "savingsEstimate": savings_estimate,
        "actionItems": action_items,
        "datasetFlags": dataset_flags,
        "ingestSource": "offline_files",
    }

    if ingest_diagnostics is not None:
        gate = evaluate_ingest_quality(ingest_diagnostics, measures["totalCost"])
        result["ingestDiagnostics"] = ingest_diagnostics
        result["ingestQuality"] = gate
        result = apply_ingest_quality_gate(result, gate)

    return result
